### PR TITLE
Add `sync::RwLock`

### PR DESCRIPTION
**Pull request by tsoutsman**
Depends on theseus-os/spin-rs#2.

### CODE REVIEW

**Comment by kevinaboos** [comment]
theseus-os/spin-rs#2 has been merged in

**Comment by tsoutsman** [comment]
~~Should be ready.~~

Nevermind.

**Comment by kevinaboos** [comment]
> ~Should be ready.~
> 
> Nevermind.

Ok lmk when you'd like another review.

**Comment by tsoutsman** [comment]
@kevinaboos should be ready.

File: libs/sync/src/lib.rs [hunk 1]
```diff
@@ -22,35 +24,59 @@ pub mod spin {
...
+        rw_lock: &'a spin::RwLock<T>,
+        data: &'a Self::RwLockData,
+    ) -> (spin::RwLockReadGuard<'a, T>, Self::Guard);
+
+    fn write_rw_lock<'a, T>(
+        rw_lock: &'a spin::RwLock<T>,
+        data: &'a Self::RwLockData,
+    ) -> (spin::RwLockWriteGuard<'a, T>, Self::Guard);
+
+    fn post_rw_lock_unlock(data: &Self::RwLockData, is_writer_or_last_reader: bool);
```
Comment (kevinaboos) [inline libs/sync/src/lib.rs]: missing docs on these functions

**Review by kevinaboos — CHANGES_REQUESTED** [review]
Sorry for the delay, i left on vacation right before you readied this PR. 

Anyway, i like everything here, nice work!  One change request -- I think it'd be a good design choice to separate out the trait representing the flavor of actions for a Mutex from the trait representing the flavor of actions for an RwLock. There are a few reasons for this:
* I think it makes sense to separate them as they are typically implemented separately in both the std lib as well as crates like `spin`. 
* Also, some concrete impls of those flavor traits may not want to (or may not be able to) offer both a Mutex and an RwLock; separating them into two traits allows one to be implemented without the other, and is also a bit clearer imo.
* `RwLock<T>` typically has different Send/Sync bounds for its type `T` than `Mutex<T>` does. I realize this isn't part of the Flavor trait API, but it still speaks to the expectation of separation between the two.
* It would give the Flavor trait a more specific name: `MutexFlavor` and `RwLockFlavor`

**Review by kevinaboos — CHANGES_REQUESTED** [review]
Sorry for the delay, i left on vacation right before you readied this PR. 

Anyway, i like everything here, nice work!  One change request -- I think it'd be a good design choice to separate out the trait representing the flavor of actions for a Mutex from the trait representing the flavor of actions for an RwLock. There are a few reasons for this:
* I think it makes sense to separate them as they are typically implemented separately in both the std lib as well as crates like `spin`. 
* Also, some concrete impls of those flavor traits may not want to (or may not be able to) offer both a Mutex and an RwLock; separating them into two traits allows one to be implemented without the other, and is also a bit clearer imo.
* `RwLock<T>` typically has different Send/Sync bounds for its type `T` than `Mutex<T>` does. I realize this isn't part of the Flavor trait API, but it still speaks to the expectation of separation between the two.
* It would give the Flavor trait a more specific name: `MutexFlavor` and `RwLockFlavor`

**Review by kevinaboos — APPROVED** [review]
awesome work!

**Pull request merged**